### ISSUE TITLE
[Feature]: Generate `build.d.ts` for server builds using TypeScript

### ISSUE THREAD

**Issue by shamsup** [remix-run/remix#1655]
### What is the new or updated feature that you are suggesting?

`remix build` should generate a corresponding `build.d.ts` (file name based on `serverBuildDirectory` from `remix.config.js`) file for remix apps that are written with TypeScript.

Example output (what I am doing in my projects based on the `create-remix` arc template):

```ts
// server/build.d.ts

import { ServerEntryModule } from '@remix-run/server-runtime';
import { AssetsManifest } from '@remix-run/server-runtime/entry';
import { ServerRouteManifest } from '@remix-run/server-runtime/routes';

declare const entry: {
  module: ServerEntryModule;
};

declare const routes: ServerRouteManifest;
declare const assets: AssetsManifest;

export { entry, routes, assets };
```

### Why should this feature be included?

Since most adapters/deployment targets require the build output from remix to be `import/require`-d in an additional file in order to be run, having the types for the build output generated would make it easier and more convenient to include the build output into a TypeScript project. 

Aside from projects using TypeScript, this also allows editors to infer the types and give meaningful code completion suggestions when working with custom deployment targets or writing a custom adapter. 

This would also prevent mistakes when upgrading remix projects, adapters, and templates with new releases of the framework. As an example, between 1.0.6 and 1.1.1, the default export of the server bundle was removed, and `import build from './build';` had to be changed to `import * as build from './build';`, which I could have caught in the editor before attempting to deploy by having the types generated by the remix compiler rather than relying on my own types.

**Comment by brophdawg11** [comment]
👋 @shamsup - I'm going to close this out as we no longer recommend importing from `./build` instead we now have the `esbuild` virtual module `@remix-run/dev/server-build` which has built-in type support.  Here's an example from a fresh `create-remix` with the Architect template:

<img width="786" alt="Screen Shot 2022-05-02 at 1 01 14 PM" src="https://user-images.githubusercontent.com/1609022/166293584-3425825b-e882-4d31-808c-dcafca1693ff.png">

Please feel free to re-open if this is still an issue for you though!

**Comment by shamsup** [comment]
Yea I think this issue was well-covered by both the removal of magic imports and allowing remix to build/bundle the `app/index` file as part of the app.

**Comment by drew-y** [comment]
Hey @brophdawg11 Is this still the recommended method of importing server builds? The docs all still say to use `import * as build from "./build/index.js`. I ran into this post trying to resolve this exact issue myself.

For context, I'm migrating a React Router app (v6) with my own express server to Remix. 

I was able to get this working by specifying the location of my `server.ts` using the `remix.config.js` `server` field. However, its unclear to me if this is actually how `@remix-run/dev/server-build` is intended to be used in the first place, since the docs don't mention it. 

Also, what would I do if I didn't want remix to own the build of my server? It seems like bundling the types would still important for external servers that consume the build.
